server/api/controllers/exam-bank-controller.ts

```diff
@@ -0,0 +1,110 @@
+import { Dirent, readdirSync } from "fs";
+import { Exam } from "../../types/exam-bank";
+import { Logger } from "../../util/logger";
+import { ReadWriteController } from "./read-write-controller";
+
+export class ExamBankController {
+  static logger = new Logger("Exam Bank Controller");
+
+  /**
+   * Re-creates the exam list JSON file based on the current 
+   * contents of the exams directory 
+   */
+  static rewriteFile(): void {
+    const url = "_hidden/exams-list";
+    ReadWriteController.overwriteJSONDataPath(
+      url,
+      (statusCode: number) => {
+        switch (statusCode) {
+          case 400:
+            this.logger.error(`Bad request made`);
+            break;
+          case 403:
+            this.logger.error(`Forbidden`);
+            break;
+          case 404:
+            this.logger.error(`${url} not found`);
+            break;
+          case 200:
+          case 201:
+            this.logger.info("Exams file rewritten");
+            break;
+        }
+      },
+      this.generateJSON()
+    );
+  }
+
+  /**
+   * Gets an array of every exam currently in the exam list folder
+   */
+  private static generateJSON(): Exam[] {
+    const examFiles: Dirent[] = readdirSync("public/exams", {
+      withFileTypes: true,
+    });
+
+    const unfilteredExams: {
+      [key: string]: Exam;
+    } = {};
+
+    for (const file of examFiles) {
+      const parts = file.name.split("-");
+      const department = parts[0],
+        course = parts[1],
+        term = parts[2],
+        type = parts.slice(3).join(" ").split(".")[0].replace(" sol", ""); // strip file extension and solutions marker
+
+      const isSolution = file.name.includes("-sol");
+      const mapKey = [department, course, term].join("-");
+
+      if (unfilteredExams[mapKey]) {
+        if (isSolution) {
+          unfilteredExams[mapKey].solutionFile = file.name;
+        } else {
+          unfilteredExams[mapKey].type = type;
+          unfilteredExams[mapKey].examFile = file.name;
+        }
+      } else {
+        unfilteredExams[mapKey] = {
+          department,
+          courseCode: course,
+          term,
+          type,
+          examFile: isSolution ? undefined : file.name,
```

Comment:
what is the logic for this line supposed to be?
if we have a solution, then the examFile for the exam should be `undefined`? this line kinda confusing me, could you explain

Reply:
The loop goes through the exam files one-by-one, constructing each exam's data structure as it goes.  
* If it reaches an exam (i.e. not a solution) file first where that exam does not exist yet in the dictionary, then it will create an entry for that exam with no solution defined yet.  
* If it reaches a solution first where it's corresponding exam does not exist yet in the dictionary, it will create an entry for that exam with no non-solution file defined yet.  
* If it reaches either an exam or solution file where there's already an entry in the dictionary, it just adds the file to that entry.

So to answer your question, let's say the loop finds a solution file for Fall 2020 Math 137's midterm.  If it hasn't seen anything from Fall 2020 Math 137, the loop will store the solution file under an entry for Math 137 Fall 2020.  Since we don't have an exam file yet, and don't know if one is coming, we just set it to undefined.

I updated it to have more comments explaining the logic!